vm/src/pyobject.rs

```diff
@@ -1599,6 +1625,33 @@ pub trait PyValue: fmt::Debug + Sized + 'static {
     }
 }
 
+pub trait PyImmutableClass: PyValue {
+    fn create_type(ctx: &PyContext) -> PyClassRef;
+
+    fn into_ref_with_context(self, ctx: &PyContext) -> PyRef<Self> {
+        PyRef {
+            obj: PyObject::new(self, Self::class_from_ctx(ctx).into_object()),
+            _payload: PhantomData,
+        }
+    }
+
+    fn class_from_ctx(ctx: &PyContext) -> PyClassRef {
+        if let Some(class) = ctx.get_type::<Self>() {
```

Comment:
I'm not sure I like this lazy initialisation of types. If we go down this path every type check will require checking that the type exists and looking it up in a hash table. A struct of builtin types seems simpler and easier.

Reply:
Agreed. I'm in favor of keeping the struct with types, like the exception zoo.